taiga2/controllers/models_controller.py

```diff
@@ -736,48 +760,140 @@ def format_renamed(datafiles):
-        new_description=new_description,
-        changes_description=changes_description,
+
+def generate_previous_version_upload_files(existing_taiga_id, sid):
+    try:
+        data_file = get_datafile_by_taiga_id(existing_taiga_id, one_or_none=True)
+    except InvalidTaigaIdFormat as ex:
+        api_error(
+            "The following was not formatted like a valid taiga ID: {}".format(
+                ex.taiga_id
+            )
+        )
+
+    if data_file is None:
+        api_error("Unknown taiga ID: " + existing_taiga_id)
+
+    return add_upload_session_virtual_file(
+        session_id=sid, filename=data_file.name, data_file_id=data_file.id
     )
 
-    activity = VersionAdditionActivity(
-        user_id=current_user.id,
-        dataset_id=new_dataset_version.dataset_id,
-        type=Activity.ActivityType.added_version,
-        dataset_description=new_description
-        if new_description != latest_dataset_version.description
-        else None,
-        dataset_version=new_dataset_version.version,
-        comments=comments,
+
+def get_previous_version_and_added_datafiles(
+    dataset_version: DatasetVersion,
+    dataset_id: str,
+    datafile_names: List[str],
+    session_id: str,
+) -> List[DataFile]:
+    version = None
+    if dataset_version is None:
+        version = get_latest_dataset_version(dataset_id)
+    else:
+        assert dataset_version.version is not None
+        version = get_dataset_version_by_dataset_id_and_version(
+            dataset_id, dataset_version.version
+        )
+
+    previous_version_datafiles = get_previous_version_upload_datafiles(
+        version.datafiles, version.dataset.permaname, version.version
     )
+
+    # print(f"LOCK STATUS LINE 657: {models_controller.is_db_locked()}")
+
+    # If file names being uploaded match an existing file from the previous
+    # version, we automatically replace the existing file with the new
+    # file.
+    previous_version_virtual_datafiles: List[UploadVirtualDataFile] = []
+    if previous_version_datafiles is not None:
+        for upload_datafile in previous_version_datafiles:
+            if upload_datafile.file_name not in datafile_names:
+                previous_version_virtual_datafiles.append(upload_datafile)
+
+    for file in previous_version_virtual_datafiles:
+        generate_previous_version_upload_files(file.taiga_id, session_id)
+
+    added_datafiles = add_datafiles_from_session(session_id)
+
+    return added_datafiles
+
+
+def is_db_locked():
+    try:
+        random_val = randint(1, 999)
+        db.session.execute(f"UPDATE lock_table SET random = {random_val}")
+        db.commit()
+        return False
+    except:
+        return True
+
+
+def lock():
+    # There was a race condition if taigapy's update_dataset was run from multiple independent
+    # processes. If add_existing_files was set to True, old files would sometimes be missing
+    # from the newer dataset versions. This lock was added to fix this.
```

Comment:
How about adding a little more explanation of how this "lock" works since it looks super suspicious.  Maybe something like `We implement a lock by relying on the database to block clients which attempt to update the same row. Once the transaction which successfully updates this row is either committed or rolled back the "lock" will be released`